### PR TITLE
Update module_utils.six to latest

### PR DESCRIPTION
**Pull request by abadger**
We've been held back on the version of six we could use on the module
side to 1.4.x because of python-2.4 compatibility.  Now that our minimum
is Python-2.6, we can update to the latest version of six in
module_utils and get rid of the second copy in lib/ansible/compat.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ansible/module_utils/six.py

##### ANSIBLE VERSION

```
devel
```

### CODE REVIEW

File: lib/ansible/plugins/action/template.py [hunk 1]
```diff
@@ -45,8 +44,8 @@ def get_checksum(self, dest, all_vars, try_directory=False, source=None, tmp=Non
                 dest = os.path.join(dest, base)
                 dest_stat = self._execute_remote_stat(dest, all_vars=all_vars, follow=False, tmp=tmp)
 
-        except AnsibleError:
-            return dict(failed=True, msg=to_native(get_exception()))
+        except AnsibleError as e:
```
Comment (alikins) [inline lib/ansible/plugins/action/template.py:47]: I think I'd remove most of syntax compatibility changes from this pr/commit, and do those in a different commit/pr.

(At least in cases where there are also non-syntax related changes that may eventually need to be back-ported/cherry-picked. Granted, in this case, the non-syntax changes are also not  2.4 compatible so it is unlikely to ever get backported to a works-with-py24 branch).

But in the past I've found it useful to have syntax related changes isolated.

File: test/units/inventory/test_host.py [hunk 2]
```diff
@@ -18,11 +18,10 @@
 # for __setstate__/__getstate__ tests
 import pickle
 
-from ansible.compat.six import string_types
-from ansible.compat.tests import unittest
-
 from ansible.inventory.group import Group
 from ansible.inventory.host import Host
+from ansible.module_utils.six import string_types
+from ansible.module_utils.tests import unittest
```
Comment (alikins) [inline test/units/inventory/test_host.py]: ```from ansible.compat.tests import unittest```
 seems to have gotten incorrectly updated to reference module_utils.tests

```
============================================================================================= ERRORS ==============================================================================================
_______________________________________________________________________ ERROR collecting test/units/inventory/test_host.py ________________________________________________________________________
ImportError while importing test module '/home/adrian/src/ansible/test/units/inventory/test_host.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../.local/lib/python2.7/site-packages/_pytest/python.py:418: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
../../.local/lib/python2.7/site-packages/py/_path/local.py:662: in pyimport
    __import__(modname)
../../.local/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:216: in load_module
    py.builtin.exec_(co, mod.__dict__)
test/units/inventory/test_host.py:24: in <module>
    from ansible.module_utils.tests import unittest
E   ImportError: No module named tests
```

File: test/units/executor/module_common/test_recursive_finder.py [hunk 3]
```diff
@@ -115,20 +115,20 @@ def test_from_import_six(self, finder_containers):
         recursive_finder(name, data, *finder_containers)
         assert finder_containers.py_module_names == set((('six',),))
         assert finder_containers.py_module_cache == {}
-        assert frozenset(finder_containers.zf.namelist()) == frozenset(('ansible/module_utils/six.py',))
+        assert frozenset(finder_containers.zf.namelist()) == frozenset(('ansible/module_utils/__init__.py', 'ansible/module_utils/_six.py'))
```
Comment (alikins) [inline test/units/executor/module_common/test_recursive_finder.py]: Not sure why, but I'm seeing unit test failures on these (py2.7 and py3.6):

```
____________________________________________________________________________ TestRecursiveFinder.test_from_import_six _____________________________________________________________________________
self = <test_recursive_finder.TestRecursiveFinder object at 0x7ff40d32c250>
finder_containers = FinderContainers(py_module_names=set([('six',)]), py_module_cache={}, zf=<zipfile.ZipFile object at 0x7ff40d33c410>)

    def test_from_import_six(self, finder_containers):
        name = 'ping'
        data = b'#!/usr/bin/python\nfrom ansible.module_utils import six'
        recursive_finder(name, data, *finder_containers)
        assert finder_containers.py_module_names == set((('six',),))
        assert finder_containers.py_module_cache == {}
>       assert frozenset(finder_containers.zf.namelist()) == frozenset(('ansible/module_utils/__init__.py', 'ansible/module_utils/_six.py'))
E       assert frozenset(['a...tils/six.py']) == frozenset(['an...ils/_six.py'])
E         Extra items in the left set:
E         'ansible/module_utils/six.py'
E         Extra items in the right set:
E         'ansible/module_utils/_six.py'
E         'ansible/module_utils/__init__.py'
E         Full diff:
E         - frozenset(['ansible/module_utils/six.py'])
E         + frozenset(['ansible/module_utils/__init__.py', 'ansible/module_utils/_six.py'])

test/units/executor/module_common/test_recursive_finder.py:118: AssertionError
_______________________________________________________________________________ TestRecursiveFinder.test_import_six _______________________________________________________________________________
self = <test_recursive_finder.TestRecursiveFinder object at 0x7ff40d33c490>
finder_containers = FinderContainers(py_module_names=set([('six',)]), py_module_cache={}, zf=<zipfile.ZipFile object at 0x7ff40d33c710>)

    def test_import_six(self, finder_containers):
        name = 'ping'
        data = b'#!/usr/bin/python\nimport ansible.module_utils.six'
        recursive_finder(name, data, *finder_containers)
        assert finder_containers.py_module_names == set((('six',),))
        assert finder_containers.py_module_cache == {}
>       assert frozenset(finder_containers.zf.namelist()) == frozenset(('ansible/module_utils/__init__.py', 'ansible/module_utils/_six.py'))
E       assert frozenset(['a...tils/six.py']) == frozenset(['an...ils/_six.py'])
E         Extra items in the left set:
E         'ansible/module_utils/six.py'
E         Extra items in the right set:
E         'ansible/module_utils/_six.py'
E         'ansible/module_utils/__init__.py'
E         Full diff:
E         - frozenset(['ansible/module_utils/six.py'])
E         + frozenset(['ansible/module_utils/__init__.py', 'ansible/module_utils/_six.py'])

test/units/executor/module_common/test_recursive_finder.py:126: AssertionError
____________________________________________________________________ TestRecursiveFinder.test_import_six_from_many_submodules _____________________________________________________________________
self = <test_recursive_finder.TestRecursiveFinder object at 0x7ff40d33c790>
finder_containers = FinderContainers(py_module_names=set([('six',)]), py_module_cache={}, zf=<zipfile.ZipFile object at 0x7ff40d33ca10>)

    def test_import_six_from_many_submodules(self, finder_containers):
        name = 'ping'
        data = b'#!/usr/bin/python\nfrom ansible.module_utils.six.moves.urllib.parse import urlparse'
        recursive_finder(name, data, *finder_containers)
        assert finder_containers.py_module_names == set((('six',),))
        assert finder_containers.py_module_cache == {}
>       assert frozenset(finder_containers.zf.namelist()) == frozenset(('ansible/module_utils/__init__.py', 'ansible/module_utils/_six.py'))
E       assert frozenset(['a...tils/six.py']) == frozenset(['an...ils/_six.py'])
E         Extra items in the left set:
E         'ansible/module_utils/six.py'
E         Extra items in the right set:
E         'ansible/module_utils/_six.py'
E         'ansible/module_utils/__init__.py'
E         Full diff:
E         - frozenset(['ansible/module_utils/six.py'])
E         + frozenset(['ansible/module_utils/__init__.py', 'ansible/module_utils/_six.py'])

test/units/executor/module_common/test_recursive_finder.py:134: AssertionError
```

**Review by alikins — CHANGES_REQUESTED** [review]
Couple of unit tests failures

**Comment by alikins** [comment]
HEADS UP. I rebased this to devel to fix the merge conflict, and fixed the test_host.py issue and updated the sanity tests to exclude six.py. 

Also changed test_recursive_finder.py to get the tests passing but if the previous expected results are supposed to include ansible/module_utils/__init__.py, that will need fixing.

(Also, pushed a copy of your branch before the rebase to my fork at https://github.com/alikins/ansible/tree/update-module_utils-six if needed for reference)

**Review by alikins — APPROVED** [review]

**Comment by mattclay** [comment]
I've overhauled the ansible-test handling of module_utils to resolve the issue that was blocking CI on this PR. Tests are now able to run, although many are still failing.

**Comment by abadger** [comment]
Merged to devel.  Not a candidate for 2.3 as it introduces incompatibility with python-2.4 on the managed machines.

**Pull request merged**